catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,605 @@
   }
 
-  public int getInterval() {
-    return interval;
+  /**
+   * Sets the {@link FederatedSource}s of which to check the availability periodically according to
+   * {@link #setPollIntervalMinutes(long)}
+   *
+   * @throws IllegalArgumentException if {@code federatedSource} is {@code null}
+   */
+  public void setFederatedSources(final List<FederatedSource> federatedSources) {
+    notNull(federatedSources);
+
+    LOGGER.trace(
+        "Setting federated sources: old size={}, new size={}",
+        this.federatedSources.size(),
+        federatedSources.size());
+    this.federatedSources = federatedSources;
   }
 
-  public void setInterval(int interval) {
-    if (interval < 1) {
-      throw new IllegalArgumentException("Cannot set to lower than 1 minute");
+  /**
+   * Sets the {@link CatalogProvider}s of which to check the availability periodically according to
+   * {@link #setPollIntervalMinutes(long)}
+   *
+   * @throws IllegalArgumentException if {@code catalogProvider} is {@code null}
+   */
+  public void setCatalogProviders(final List<CatalogProvider> catalogProviders) {
+    notNull(catalogProviders);
+
+    LOGGER.trace(
+        "Setting catalog providers: old size={}, new size={}",
+        this.catalogProviders.size(),
+        catalogProviders.size());
+    this.catalogProviders = catalogProviders;
+  }
+
+  /**
+   * Sets the {@link CatalogStore}s of which to check the availability periodically according to
+   * {@link #setPollIntervalMinutes(long)}
+   *
+   * @throws IllegalArgumentException if {@code catalogStore} is {@code null}
+   */
+  public void setCatalogStores(final List<CatalogStore> catalogStores) {
+    notNull(catalogStores);
+
+    LOGGER.trace(
+        "Setting catalog stores: old size={}, new size={}",
+        this.catalogStores.size(),
+        catalogStores.size());
+    this.catalogStores = catalogStores;
+  }
+
+  /** @throws IllegalArgumentException if one of the sources has not already been bound */
+  private void recheckAvailabilityOfAllSourcesAndWait() {
+    final Set<Source> sources =
+        Stream.of(connectedSources, federatedSources, catalogProviders, catalogStores)
+            .flatMap(Collection::stream)
+            .collect(Collectors.toSet());
+
+    LOGGER.trace(
+        "Found {} sources. Starting a process to check the availability of each source",
+        sources.size());
+
+    // TODO Unbind orphaned entries in the map caused by when sources are modified and their key is
+    // different
+
+    final Set<Future> futures =
+        sources
+            .stream()
+            .map(
+                source -> {
+                  final String sourceId = source.getId();
+                  final SourceKey sourceKey = new SourceKey(source);
+
+                  final Optional<CachedSourceAvailability> cachedSourceStatusOptional =
+                      Optional.ofNullable(cache.getIfPresent(sourceKey));
+                  if (cachedSourceStatusOptional.isPresent()) {
```

Comment:
I prefer Optionals over nullable values. Regarding using the loader, I like explicitly checking so that I can different message in either of the cases.
```java
final Optional<CachedSourceAvailability> cachedSourceAvailabilityOptional =
    Optional.ofNullable(cache.getIfPresent(sourceKey));
if (cachedSourceAvailabilityOptional.isPresent()) {
  final CachedSourceAvailability cachedSourceAvailability =
      cachedSourceAvailabilityOptional.get();
  LOGGER.trace(
      "Starting a process to recheck the availability of source id={}, which has already been added to the cache ",
      source.getId());
  cachedSourceAvailability.recheckAvailability(
      source, pollInterval, pollIntervalTimeUnit, sourceAvailabilityChecksThreadPool);
} else {
  LOGGER.debug("Found a new SourceKey for source id={}", source.getId());
  final CachedSourceAvailability cachedSourceAvailability = new CachedSourceAvailability();
  cache.put(sourceKey, cachedSourceAvailability);
  cachedSourceAvailability.recheckAvailability(
      source, pollInterval, pollIntervalTimeUnit, sourceAvailabilityChecksThreadPool);
}
```